### PR TITLE
Fixes #29939 - settings in memory

### PR DESCRIPTION
**Pull request by ezr-ondrej**
This came to live as an effort to make extending settings in category bit more easier.
It developed in a draft of:
- keep only changed settings in database
- keep all the settings data in memory

Let me know what do you think on [discourse](https://community.theforeman.org/t/setting-in-memory-only-changed-values-in-database/18976)

### CODE REVIEW

**Comment by lzap** [comment]
```
TypeError: no implicit conversion of String into Array
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:100:in `|'
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:100:in `add_default'
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/models/setting.rb:271:in `set'
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/models/setting/auth.rb:4:in `default_settings'
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:121:in `each'
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:121:in `load_defaults'
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:95:in `load'
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/config/application.rb:332:in `block in <class:Application>'
```

**Comment by ezr-ondrej** [comment]
> ```
> TypeError: no implicit conversion of String into Array
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:100:in `|'
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:100:in `add_default'
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/models/setting.rb:271:in `set'
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/models/setting/auth.rb:4:in `default_settings'
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:121:in `each'
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:121:in `load_defaults'
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:95:in `load'
> /home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/config/application.rb:332:in `block in <class:Application>'
> ```

Thanks, fixed :)

**Comment by kamils-iRonin** [comment]
```
NoMethodError: undefined method `categories=' for #<Foreman::SettingManager:0x000000000a9d2da8 @settings={}>
Did you mean?  categories
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.5/slave/fast/app/registries/foreman/setting_manager.rb:100:in `add_default'
```

File: app/registries/foreman/setting_manager.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      Setting.descendants.each(&:default_settings)
+    end
+
+    def load_settings(category = nil)
+      return false unless Setting.table_exists?
+      Setting.where(:category => category || categories).each do |set|
+        if settings.key?(set.name)
+          settings[set.name].value = set.value
+        else
+          next if Rails.env.development?
```
Comment (ekohl) [inline app/registries/foreman/setting_manager.rb]: I'd at least expect a warning about this in dev

File: app/registries/foreman/setting_manager.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      end
+
+      def readonly?
+        SETTINGS.key?(name.to_sym)
+      end
+
+      def has_default?
+        default_type = settings_type || default.class.to_s.downcase
+        case default_type
+        when "array", "hash", "string"
```
Comment (domitea) [inline app/registries/foreman/setting_manager.rb]: I found that `%w(array hash string)` is more readable but not strong opinion here

It's better to use single quotes in hat case because there is no string interpolation

File: config/application.rb [hunk 3]
```diff
@@ -326,6 +326,12 @@ class Application < Rails::Application
       routes_reloader.execute_if_updated
     end
 
+    config.after_initialize do
```
Comment (domitea) [inline config/application.rb:330]: Is there a reason call again `after_initialize` ?

File: lib/foreman.rb [hunk 4]
```diff
@@ -30,4 +30,8 @@ def self.instance_id
   def self.instance_id=(value)
     Setting[:instance_id] = value
   end
+
+  def self.setting_manager
+    @setting_manager ||= Foreman::SettingManager.new
```
Comment (domitea) [inline lib/foreman.rb:35]: Nice singleton!

File: app/models/setting.rb [hunk 5]
```diff
@@ -87,22 +86,19 @@ def self.humanized_category
...
-    record = where(:name => name).first!
-    record.value = value
-    record.save!
+    Foreman.setting_manager.update_setting(name, value)
+  end
+
+  def self.setting_type_from_value(default)
+    t = default.class.to_s.downcase
+    return t if TYPES.include?(t)
+    return "integer" if default.is_a?(Integer)
```
Comment (domitea) [inline app/models/setting.rb:104]: I don't think that this condition is necessary because integer is taken by first one.

`5.class.to_s.downcase => "integer"`

And that brings me a question - How we can handle another numeric types ie. float or decimal?

File: config/application.rb (hunk 3 above)
Comment (ezr-ondrej) [reply config/application.rb:330]: It serves a different purpose :man_shrugging:

File: app/models/setting.rb (hunk 5 above)
Comment (ezr-ondrej) [reply app/models/setting.rb:104]: Not true, It can be and most likely will be `Fixnum`, that is sub-class of Integer, but `<fixnum>.class.to_s.downcase` won't help :)

File: app/registries/foreman/setting_manager.rb (hunk 2 above)
Comment (ezr-ondrej) [reply app/registries/foreman/setting_manager.rb]: we would need to do *%w(array hash string) and that's not as readable I'd say.
But :+1: for single quotes.

**Comment by ezr-ondrej** [comment]
[test foreman]

File: app/helpers/settings_helper.rb [hunk 6]
```diff
@@ -1,31 +1,35 @@
 module SettingsHelper
-  def value(setting)
-    if setting.readonly?
-      return readonly_field(setting, :value,
-        {:title => _("This setting is defined in the configuration file '%{filename}' and is read-only.") % {:filename => setting.class.config_file}, :helper => :show_value})
-    end
+  def setting_presenter(setting)
+    presenter = Foreman.setting_manager.settings[setting.name]
```
Comment (xprazak2) [inline app/helpers/settings_helper.rb:3]: What is the preferred way to get the setting we need? `Foreman.setting_manager.settings[setting.name]` or `Setting[setting.name]`? I think the former bypasses the cache...

File: app/services/classification/values_hash_query.rb [hunk 7]
```diff
@@ -69,11 +69,11 @@ def split_matcher(lookup_value)
         matcher_key = ''
         matcher_value = ''
 
-        lookup_value.match.split(LookupKey::KEY_DELM).each do |match_key|
-          element = match_key.split(LookupKey::EQ_DELM).first
-          matcher_key += element + ','
-          if LookupKey::MATCHERS_INHERITANCE.include?(element)
-            matcher_value = match_key.split(LookupKey::EQ_DELM).last
+        lookup_value.match.split(LookupKey::KEY_DELM).each do |match_keyval|
```
Comment (xprazak2) [inline app/services/classification/values_hash_query.rb]: Are these changes related?

File: test/unit/classification_test.rb [hunk 8]
```diff
@@ -540,421 +540,389 @@ def setup
...
-    as_admin do
-      LookupValue.create! :lookup_key_id => key.id,
-                          :match => "hostgroup=#{parent_hostgroup}",
-                          :value => ['parent'],
-                          :omit => false
-      LookupValue.create! :lookup_key_id => key.id,
-                          :match => "organization=#{taxonomies(:organization1)}",
-                          :value => ['org'],
-                          :omit => false
+  context 'with matchers_inheritance enabled' do
```
Comment (xprazak2) [inline test/unit/classification_test.rb]: Are changes in this test file related?

**Comment by xprazak2** [comment]
I am getting a bunch of warnings from the plugins: 

settings-not-defined

File: app/registries/foreman/setting_manager.rb [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    attribute :encrypted, :boolean, :default => false
+    attribute :settings_type, :string
+    attr_accessor :preset_value, :options
+
+    def self.model_name
+      Setting.model_name
+    end
+
+    def model_name
+      Setting.model_name
```
Comment (xprazak2) [inline app/registries/foreman/setting_manager.rb:35]: `self.class.model_name` to stay dry?

File: app/helpers/settings_helper.rb (hunk 6 above)
Comment (ezr-ondrej) [reply app/helpers/settings_helper.rb:3]: No, it does not bypass the cache. I've moved the cache to the `SettingManager` so no they are equal, I'd probably keep `Setting[setting.name]` as preferred way to get setting value and that is just a proxy for `settin_manager[setting.name]`. But `Foreman.setting_manager.settings[setting.name]` is accessing the setting registry, where we keep all info about the setting and returns the `Presenter`.
It's probably bit confusing and we can drop `SettingManager#[]` and call `SettingManager#read_setting` from `Setting[]` directly.

File: app/services/classification/values_hash_query.rb (hunk 7 above)
Comment (ezr-ondrej) [reply app/services/classification/values_hash_query.rb]: no, they are probably not, the tests for them just started to fail with my changes. I forgot I've not split them.
Going to do so now ^_^

File: test/unit/classification_test.rb (hunk 8 above)
Comment (ezr-ondrej) [reply test/unit/classification_test.rb]: yup, they were, but those can be done prior this change. Let me drop these

**Comment by ezr-ondrej** [comment]
> I am getting a bunch of warnings from the plugins:

If I've implemented the warning correctly, that means you've got mess in your database (e.g. settings that were introduced in a PR you've tested, but ended up having different name, settings that were dropped, but didn't have migration to be erased from DB). I'd like to introduce a rake/startup check to clean them up.

File: app/models/setting.rb [hunk 10]
```diff
@@ -264,42 +261,19 @@ def self.bypass_readonly(s, &block)
...
-        val = s.update(:category => name).symbolize_keys
-        dbcache.key?(val[:name]) ? create_existing(dbcache[val[:name]], s) : create!(s)
-      end
-    end
-    true
-  end
-
-  def self.select_collection_registry
-    @@select_collection ||= SettingSelectCollection.new
+    nil # Already loaded
```
Comment (xprazak2) [inline app/models/setting.rb:271]: If settings are already loaded, do we still need this method? I do not see it being called anywhere...

**Comment by xprazak2** [comment]
Regarding the warnings, they are caused by [foreman_templates](https://github.com/theforeman/foreman_templates/blob/master/app/models/setting/template_sync.rb#L69) having `load_defaults` instead of `default_settings`. 

Also no settings for katello are shown.

File: app/models/setting.rb (hunk 10 above)
Comment (ezr-ondrej) [reply app/models/setting.rb:271]: `Katello` is calling it explicitly. But we should probably deprecate it :thinking:
Let me take a look into it.

**Comment by ares** [comment]
As discussed with @ezr-ondrej, this is replaced with smaller PRs, closing this. Thanks for the effort so far, keep it going!
